Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [ ]:
NAME = "Ivan Dario Ovalle Benavides"
COLLABORATORS = ""

---

<img src="https://hadoop.apache.org/docs/r1.2.1/images/hadoop-logo.jpg">

# Sistema de ficheros HDFS: configuración del entorno de la asignatura

Como ya se ha visto en la parte de teoría el sistema de archivos Hadoop (HDFS) es una parte fundamental del entorno Big Data de Apache Hadoop. En esta práctica primeramente exploraremos cómo podemos interactuar desde la línea de comandos con el sistema de ficheros HDFS. El primer paso es abrir un terminal desde el Jupyter o bien ejecutar los comandos mediante el comando "!" dentro del Jupyter.

Los comandos que se pueden enviar al sistema de archivos son muy similares a las de bash en entornos Linux.

Para listar los archivos del directorio raíz del HDFS utilizaremos el comando *ls*. 

`hdfs dfs -ls /`

In [1]:
!hdfs dfs -ls /

Found 9 items
drwxr-xr-x   - asolerib   supergroup          0 2020-01-16 22:12 /CFCC
drwxr-xr-x   - asolerib   supergroup          0 2021-05-13 17:37 /aula_B0.476
drwxr-xr-x   - asolerib   supergroup          0 2019-10-28 11:09 /aula_B2.585
drwxr-xr-x   - asolerib   supergroup          0 2019-09-25 22:10 /aula_M2.858
drwxr-xr-x   - hbase      hbase               0 2021-09-13 14:37 /hbase
drwxr-xr-x   - aperezgari supergroup          0 2021-08-05 12:58 /home
drwxrwxr-x   - solr       solr                0 2019-07-23 15:49 /solr
drwxrwxrwt   - hdfs       supergroup          0 2021-05-25 22:52 /tmp
drwxrwxr-x   - hdfs       supergroup          0 2021-09-27 10:19 /user


0. Ahora se pide listar los ficheros del directorio /aula_M2.858. (1 punto)

Se pide rellenar la siguiente celda con el comando requerido.

In [2]:
!hdfs dfs -ls  /aula_M2.858

Found 2 items
drwxr-xr-x   - asolerib supergroup          0 2020-09-21 22:02 /aula_M2.858/PECs
drwxr-xr-x   - asolerib supergroup          0 2021-10-06 16:09 /aula_M2.858/data


Tal y como se verá a continuación los comandos que HDFS proporciona, permiten: listar, consultar, subir o bajar archivos desde el directorio local en HDFS. Como ya se ha podido ver el directorio `aula_M2.858` contiene una serie de subdirectorios donde se proporcionarán algunos archivos para hacer los siguientes ejercicios. A continuación se pide ejecutar una serie de operaciones que dejarán su directorio hdfs personal configurado para la realización de la asignatura. Puede encontrar el manual de los comandos del sistema HDFS en el siguiente link [HDFS Commands](https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-hdfs/HDFSCommands.html).



1. Se pide crear la carpeta "data" dentro del directorio HDFS /user/\< loginestudiante \>/

In [3]:
!hdfs dfs -mkdir /user/ivanovalle/data

mkdir: `/user/ivanovalle/data': File exists


2. Copiar al directorio "data" que acabáis de crear el fichero "Folds5x2_pp.csv" que se encuentra en el directorio HDFS /aula_M2.858/data/. Utilizad el comando `hdfs dfs -cp` para copiar el fichero.

In [3]:
!hdfs dfs -cp /aula_M2.858/data/Folds5x2_pp.csv /user/ivanovalle/data/Folds5x2_pp.csv

cp: `/user/ivanovalle/data/Folds5x2_pp.csv': File exists


3. Copiar las 10 primeras líneas del fichero LoremIpsum.txt que se encuentra en el directorio HDFS /aula_M2.858/data/ al fichero LoremIpsum_10.txt al directorio HDFS /user/\< loginestudiante \>/data. Os serán de utilidad los comandos `hdfs dfs -cat` i `hdfs dfs -put` a parte del comando clásico de bash `head`.

In [6]:
#START <FILL IN>
!hdfs dfs -cat /aula_M2.858/data/LoremIpsum.txt | head -n 10 > LoremIpsum_10.txt
!hdfs dfs -put LoremIpsum_10.txt /user/ivanovalle/data/LoremIpsum_10.txt
#END <FILL IN>

cat: Unable to write to output stream.
put: `/user/ivanovalle/data/LoremIpsum_10.txt': File exists


A parte de los comandos que ya habéis visto orientados a que cada usuario pueda gestionar el sistema de ficheros HDFS, también es posible explorar la organización interna de los diferentes ficheros existentes en los directorios HDFS.

4. Se pide que mediante el comando `hdfs fsck` determinar y describir como esta estructurado el fichero _/aula_M2.858/data/LoremIpsum.txt_ dentro del sistema HDFS. En concreto se pide mostrar como está subdividido en bloques, en que nodo están los bloques y si estos están replicados. Podéis consultar como utilizar el comando `hdfs fsck` en la web oficial de referéncia -> [HDFS Commands](https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-hdfs/HDFSCommands.html)

In [18]:
#Run the command to obtain the required information regarding the structure of file `/aula_M2.858/data/LoremIpsum.txt`
#START <FILL IN> 
!hdfs fsck /aula_M2.858/data/LoremIpsum.txt -files -blocks -racks
#END <FILL IN>

Connecting to namenode via http://eimtcld.uoc.edu:9870/fsck?ugi=ivanovalle&files=1&blocks=1&racks=1&path=%2Faula_M2.858%2Fdata%2FLoremIpsum.txt
FSCK started by ivanovalle (auth:SIMPLE) from /213.73.35.119 for path /aula_M2.858/data/LoremIpsum.txt at Sun Oct 10 06:55:03 CEST 2021
/aula_M2.858/data/LoremIpsum.txt 203431368 bytes, replicated: replication=3, 2 block(s):  OK
0. BP-2074018746-213.73.35.119-1563889676427:blk_1073781405_40591 len=134217728 Live_repl=3  [/default/10.20.30.3:9866, /default/10.20.30.2:9866, /default/213.73.35.119:9866]
1. BP-2074018746-213.73.35.119-1563889676427:blk_1073781406_40592 len=69213640 Live_repl=3  [/default/10.20.30.3:9866, /default/10.20.30.2:9866, /default/213.73.35.119:9866]


Status: HEALTHY
 Number of data-nodes:	3
 Number of racks:		1
 Total dirs:			0
 Total symlinks:		0

Replicated Blocks:
 Total size:	203431368 B
 Total files:	1
 Total blocks (validated):	2 (avg. block size 101715684 B)
 Minimally replicated blocks:	2 (100.0 %)
 Over-replicate

In [23]:
#Describe in words the questions regarding the file `/aula_M2.858/data/LoremIpsum.txt`

# El archivo se dividio en 2 bloques, se crearon 3 nodos de datos y 3 replicas.

# Imagino que se vería así:
# Archivo = [B1,B2]

# Cloudera01:9866 (213.73.35.119:9866) = [B1,B2]
# Cloudera02:9866 (10.20.30.2:9866) = [B1,B2]
# Cloudera03:9866 (10.20.30.3:9866) = [B1,B2]

# En donde los cloudera serían los nodos de datos, B1 y B2 los bloques, y hay 3 replicas de estos bloques.


#END <FILL IN>

5. Como sabeis los diferentes ficheros HDFS están particionados en bloques. Podeis consultar la configuración mediante el comando: `hdfs getconf -confKey dfs.blocksize`. ¿Cuantos MB hay por bloque?

In [1]:
#START <FILL IN>

!hdfs getconf -confKey dfs.blocksize

# 134217728, 128 MB

#END <FILL IN>

134217728


6. Se pide explorar la estructura del sistema HDFS mediante su WebUI. Para ello primeramente debemos creare un túnel ssh que nos dé acceso al puerto 9870 (sobre el que está corriendo la Web UI de HDFS). A los alumnos que tengáis un sistema basado en Linux, podéis hacerlo de manera muy simple mediante el siguiente comando:

```python 
ssh -p55000  -L 9870:eimtcld.uoc.edu:9870 asolerib@eimtcld.uoc.edu
        
```
Una vez ejecutado el comando y logeado correctamente podéis acceder a la Web UI de HDFS en la dirección:

```Python
http://localhost:9870/dfshealth.html#tab-overview
```

> Nota: a los alumnos que solo dispongan de un sistema Windows, adjuntamos una pequeña descripción para crear el túnel ssh mediante la aplicación Putty.

Esta pregunta es de texto libre y se pide que se describa a continuación qué información podéis obtener explorando la Web UI, número de DataNodes, localización de estos y su estado, factores de replicación, errores si los hay, etc.

In [ ]:

# Hay 3 DataNodes:
# Cloudera01:9866 (213.73.35.119:9866)
# Cloudera02:9866 (10.20.30.2:9866)
# Cloudera03:9866 (10.20.30.3:9866)

# Todos en OK (In service)

# Configured Capacity 2.18 TB, DFS Used 880.81 (39.48%)

# Default replication factor: 3

# Sin errores